setwd("~/Documents/INET-work/virus_network/figure_results/")
library(msa)
library(seqinr)
library(ape)
library(phytools)
# http://www.phytools.org/Cordoba2017/ex/2/Intro-to-phylogenies.html

nl63.pep <- readAAStringSet("../text_results/HCoV-NL63_peptide.fa")
e229.pep <- readAAStringSet("../text_results/HCoV-229E_peptide.fa")
hku1.pep <- readAAStringSet("../text_results/HCoV-HKU1_peptide.fa")
oc43.pep <- readAAStringSet("../text_results/HCoV-OC43_peptide.fa")
sars2.pep <- readAAStringSet("../text_results/SARS-CoV-2_peptide.fa")
sars.pep <- readAAStringSet("../text_results/SARS-CoV_peptide.fa")
mers.pep <- readAAStringSet("../text_results/MERS-CoV_peptide.fa")

pp1ab <- readAAStringSet("../text_results/pp1ab_pep.fa")

nl63.genome <- readDNAStringSet("../sources/HcoV-NL63/GCF_000853865.1_ViralProj14960_genomic.fna")
e229.genome <- readDNAStringSet("../sources/HcoV-229E/GCF_000853505.1_ViralProj14913_genomic.fna")
hku1.genome <- readDNAStringSet("../sources/HcoV-HKU1/GCF_000858765.1_ViralProj15139_genomic.fna")
oc43.genome <- readDNAStringSet("../sources/HcoV-OC43/GCF_003972325.1_ASM397232v1_genomic.fna")
sars2.genome <- readDNAStringSet("../sources/SARS-CoV-2/GCF_009858895.2_ASM985889v3_genomic.fna")
sars.genome <- readDNAStringSet("../sources/SARS-CoV/SARS-CoV_genome.fa")
mers.genome <- readDNAStringSet("../sources/MERS-CoV/GCF_000901155.1_ViralProj183710_genomic.fna")

nsp1 <- c(nl63.pep[1], e229.pep[1], hku1.pep[1], oc43.pep[1], sars2.pep[1], mers.pep[1], sars.pep[1])
nsp1.msa <- msa(nsp1, "ClustalW")
msaPrettyPrint(nsp1.msa, output="pdf", file = "nsp1_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp1.msa.seqinr <- msaConvert(nsp1.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp1.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp1_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP1 peptide")
dev.off()

nsp2 <- c(nl63.pep[2], e229.pep[2], hku1.pep[2], oc43.pep[2], sars2.pep[2], mers.pep[2], sars.pep[2])
nsp2.msa <- msa(nsp2, "ClustalW")
msaPrettyPrint(nsp2.msa, output="pdf", file = "nsp2_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp2.msa.seqinr <- msaConvert(nsp2.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp2.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp2_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP2 peptide")
dev.off()

nsp3 <- c(nl63.pep[3], e229.pep[3], hku1.pep[3], oc43.pep[3], sars2.pep[3], mers.pep[3], sars.pep[3])
nsp3.msa <- msa(nsp3, "ClustalW")
msaPrettyPrint(nsp3.msa, output="pdf", file = "nsp3_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp3.msa.seqinr <- msaConvert(nsp3.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp3.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp3_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP3 peptide")
dev.off()

nsp4 <- c(nl63.pep[4], e229.pep[4], hku1.pep[4], oc43.pep[4], sars2.pep[4], mers.pep[4], sars.pep[4])
nsp4.msa <- msa(nsp4, "ClustalW")
msaPrettyPrint(nsp4.msa, output="pdf", file = "nsp4_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp4.msa.seqinr <- msaConvert(nsp4.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp4.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp4_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP4 peptide")
dev.off()

nsp5 <- c(nl63.pep[5], e229.pep[5], hku1.pep[5], oc43.pep[5], sars2.pep[5], mers.pep[5], sars.pep[5])
nsp5.msa <- msa(nsp5, "ClustalW")
msaPrettyPrint(nsp5.msa, output="pdf", file = "nsp5_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp5.msa.seqinr <- msaConvert(nsp5.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp5.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp5_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP5 peptide")
dev.off()

nsp6 <- c(nl63.pep[6], e229.pep[6], hku1.pep[6], oc43.pep[6], sars2.pep[6], mers.pep[6], sars.pep[6])
nsp6.msa <- msa(nsp6, "ClustalW")
msaPrettyPrint(nsp6.msa, output="pdf", file = "nsp6_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp6.msa.seqinr <- msaConvert(nsp6.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp6.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp6_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP6 peptide")
dev.off()

nsp7 <- c(nl63.pep[7], e229.pep[7], hku1.pep[7], oc43.pep[7], sars2.pep[7], mers.pep[7], sars.pep[7])
nsp7.msa <- msa(nsp7, "ClustalW")
msaPrettyPrint(nsp7.msa, output="pdf", file = "nsp7_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp7.msa.seqinr <- msaConvert(nsp7.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp7.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp7_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP7 peptide")
dev.off()

nsp8 <- c(nl63.pep[8], e229.pep[8], hku1.pep[8], oc43.pep[8], sars2.pep[8], mers.pep[8], sars.pep[8])
nsp8.msa <- msa(nsp8, "ClustalW")
msaPrettyPrint(nsp8.msa, output="pdf", file = "nsp8_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp8.msa.seqinr <- msaConvert(nsp8.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp8.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp8_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP8 peptide")
dev.off()

nsp9 <- c(nl63.pep[9], e229.pep[9], hku1.pep[9], oc43.pep[9], sars2.pep[9], mers.pep[9], sars.pep[9])
nsp9.msa <- msa(nsp9, "ClustalW")
msaPrettyPrint(nsp9.msa, output="pdf", file = "nsp9_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp9.msa.seqinr <- msaConvert(nsp9.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp9.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp9_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP9 peptide")
dev.off()

nsp10 <- c(nl63.pep[10], e229.pep[10], hku1.pep[10], oc43.pep[10], sars2.pep[10], mers.pep[10], sars.pep[10])
nsp10.msa <- msa(nsp10, "ClustalW")
msaPrettyPrint(nsp10.msa, output="pdf", file = "nsp10_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp10.msa.seqinr <- msaConvert(nsp10.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp10.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp10_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP10 peptide")
dev.off()

nsp12 <- c(nl63.pep[11], e229.pep[11], hku1.pep[11], oc43.pep[11], sars2.pep[11], mers.pep[11], sars.pep[11])
nsp12.msa <- msa(nsp12, "ClustalW")
msaPrettyPrint(nsp12.msa, output="pdf", file = "nsp12_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp12.msa.seqinr <- msaConvert(nsp12.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp12.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp12_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP12 peptide")
dev.off()

nsp13 <- c(nl63.pep[12], e229.pep[12], hku1.pep[12], oc43.pep[12], sars2.pep[12], mers.pep[12], sars.pep[12])
nsp13.msa <- msa(nsp13, "ClustalW")
msaPrettyPrint(nsp13.msa, output="pdf", file = "nsp13_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp13.msa.seqinr <- msaConvert(nsp13.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp13.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp13_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP13 peptide")
dev.off()

nsp14 <- c(nl63.pep[13], e229.pep[13], hku1.pep[13], oc43.pep[13], sars2.pep[13], mers.pep[13], sars.pep[13])
nsp14.msa <- msa(nsp14, "ClustalW")
msaPrettyPrint(nsp14.msa, output="pdf", file = "nsp14_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp14.msa.seqinr <- msaConvert(nsp14.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp14.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp14_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP14 peptide")
dev.off()

nsp15 <- c(nl63.pep[14], e229.pep[14], hku1.pep[14], oc43.pep[14], sars2.pep[14], mers.pep[14], sars.pep[14])
nsp15.msa <- msa(nsp15, "ClustalW")
msaPrettyPrint(nsp15.msa, output="pdf", file = "nsp15_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp15.msa.seqinr <- msaConvert(nsp15.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp15.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp15_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP15 peptide")
dev.off()

nsp16 <- c(nl63.pep[15], e229.pep[15], hku1.pep[15], oc43.pep[15], sars2.pep[15], mers.pep[15], sars.pep[15])
nsp16.msa <- msa(nsp16, "ClustalW")
msaPrettyPrint(nsp16.msa, output="pdf", file = "nsp16_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp16.msa.seqinr <- msaConvert(nsp16.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp16.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp16_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP16 peptide")
dev.off()

S.protein <- c(nl63.pep[16], e229.pep[16], hku1.pep[17], oc43.pep[18], sars2.pep[16], mers.pep[16], sars.pep[16])
S.protein.msa <- msa(S.protein, "ClustalW")
msaPrettyPrint(S.protein.msa, output="pdf", file = "S.protein_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
S.protein.msa.seqinr <- msaConvert(S.protein.msa, type = "seqinr::alignment")
d <- dist.alignment(S.protein.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("S.protein_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of Spike protein peptide")
dev.off()

E.protein <- c(nl63.pep[18], e229.pep[19], hku1.pep[19], oc43.pep[19], sars2.pep[19], mers.pep[21], sars.pep[19])
E.protein.msa <- msa(E.protein, "ClustalW")
msaPrettyPrint(E.protein.msa, output="pdf", file = "E.protein_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
E.protein.msa.seqinr <- msaConvert(E.protein.msa, type = "seqinr::alignment")
d <- dist.alignment(E.protein.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("E.protein_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of Envelope protein peptide")
dev.off()

M.protein <- c(nl63.pep[19], e229.pep[20], hku1.pep[20], oc43.pep[20], sars2.pep[20], mers.pep[22], sars.pep[20])
M.protein.msa <- msa(M.protein, "ClustalW")
msaPrettyPrint(M.protein.msa, output="pdf", file = "M.protein_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
M.protein.msa.seqinr <- msaConvert(M.protein.msa, type = "seqinr::alignment")
d <- dist.alignment(M.protein.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("M.protein_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of Envelope protein peptide")
dev.off()

N.protein <- c(nl63.pep[20], e229.pep[21], hku1.pep[21], oc43.pep[21], sars2.pep[26], mers.pep[23], sars.pep[26])
N.protein.msa <- msa(N.protein, "ClustalW")
msaPrettyPrint(N.protein.msa, output="pdf", file = "N.protein_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
N.protein.msa.seqinr <- msaConvert(N.protein.msa, type = "seqinr::alignment")
d <- dist.alignment(N.protein.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("N.protein_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of Nucleocapsid protein peptide")
dev.off()

pp1ab.msa <- msa(pp1ab)
# msaPrettyPrint(pp1ab.msa, output="pdf", file = "pp1ab_pep.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
pp1ab.msa.seqinr <- msaConvert(pp1ab.msa, type = "seqinr::alignment")
d <- dist.alignment(pp1ab.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("pp1ab_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of pp1ab")
dev.off()

genome.msa <- msa(c(nl63.genome, hku1.genome, e229.genome, oc43.genome, sars.genome, sars2.genome, mers.genome))
# msaPrettyPrint(genome.msa, output="pdf", file = "genome_msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
genome.msa.seqinr <- msaConvert(genome.msa, type = "seqinr::alignment")
d <- dist.alignment(genome.msa.seqinr, "identity")
hemoTree <- nj(d)
plotTree(hemoTree, node.numbers = T)
rt.9 <- rotate(hemoTree, 9)
pdf("genome_tree.pdf")
plotTree(rt.9, main = "Phylogenetic tree of genome")
dev.off()

###############################################################################
# separate group 1 and 2; or call alpha-coronavirus and beta-coronavirus
###############################################################################
nsp1.alpha <- c(nl63.pep[1], e229.pep[1])
nsp1.beta <- c(hku1.pep[1], oc43.pep[1], sars2.pep[1], mers.pep[1], sars.pep[1])
nsp1.alpha.msa <- msa(nsp1.alpha, "ClustalW")
msaPrettyPrint(nsp1.alpha.msa, output="pdf", file = "nsp1_pep.alpha.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp1.alpha.msa.seqinr <- msaConvert(nsp1.alpha.msa, type = "seqinr::alignment")
## two proteins can't have phylogenetic tree, which is of alpha-coronavirus
nsp1.beta.msa <- msa(nsp1.beta, "ClustalW")
msaPrettyPrint(nsp1.beta.msa, output="pdf", file = "nsp1_pep.beta.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp1.beta.msa.seqinr <- msaConvert(nsp1.beta.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp1.beta.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp1.beta_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP1 peptide (beta-coronavirus)")
dev.off()

pid <- pairwiseAlignment(hku1.pep[[1]][, 1], oc43.pep[[1]][, 1], sars2.pep[[1]][, 1], mers.pep[[1]][, 1], sars.pep[[1]][, 1])

nsp3.alpha <- c(nl63.pep[3], e229.pep[3])
nsp3.beta <- c(hku1.pep[3], oc43.pep[3], sars2.pep[3], mers.pep[3], sars.pep[3])
nsp3.alpha.msa <- msa(nsp3.alpha, "ClustalW")
msaPrettyPrint(nsp3.alpha.msa, output="pdf", file = "nsp3_pep.alpha.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp3.alpha.msa.seqinr <- msaConvert(nsp3.alpha.msa, type = "seqinr::alignment")
## two proteins can't have phylogenetic tree, which is of alpha-coronavirus
nsp3.beta.msa <- msa(nsp3.beta, "ClustalW")
msaPrettyPrint(nsp3.beta.msa, output="pdf", file = "nsp3_pep.beta.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp3.beta.msa.seqinr <- msaConvert(nsp3.beta.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp3.beta.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp3.beta_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP3 peptide (beta-coronavirus)")
dev.off()

nsp5.alpha <- c(nl63.pep[5], e229.pep[5])
nsp5.beta <- c(hku1.pep[5], oc43.pep[5], sars2.pep[5], mers.pep[5], sars.pep[5])
nsp5.alpha.msa <- msa(nsp5.alpha, "ClustalW")
msaPrettyPrint(nsp5.alpha.msa, output="pdf", file = "nsp5_pep.alpha.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp5.alpha.msa.seqinr <- msaConvert(nsp5.alpha.msa, type = "seqinr::alignment")
## two proteins can't have phylogenetic tree, which is of alpha-coronavirus
nsp5.beta.msa <- msa(nsp5.beta, "ClustalW")
msaPrettyPrint(nsp5.beta.msa, output="pdf", file = "nsp5_pep.beta.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp5.beta.msa.seqinr <- msaConvert(nsp5.beta.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp5.beta.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp5.beta_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP5 peptide (beta-coronavirus)")
dev.off()

nsp12.alpha <- c(nl63.pep[11], e229.pep[11])
nsp12.beta <- c(hku1.pep[11], oc43.pep[11], sars2.pep[11], mers.pep[11], sars.pep[11])
nsp12.alpha.msa <- msa(nsp12.alpha, "ClustalW")
msaPrettyPrint(nsp12.alpha.msa, output="pdf", file = "nsp12_pep.alpha.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp12.alpha.msa.seqinr <- msaConvert(nsp12.alpha.msa, type = "seqinr::alignment")
## two proteins can't have phylogenetic tree, which is of alpha-coronavirus
nsp12.beta.msa <- msa(nsp12.beta, "ClustalW")
msaPrettyPrint(nsp12.beta.msa, output="pdf", file = "nsp12_pep.beta.msa.pdf", showLogo="none", verbose=FALSE, askForOverwrite = FALSE, showLegend = FALSE, showConsensus = "none")
nsp12.beta.msa.seqinr <- msaConvert(nsp12.beta.msa, type = "seqinr::alignment")
d <- dist.alignment(nsp12.beta.msa.seqinr, "identity")
hemoTree <- nj(d)
pdf("nsp12.beta_tree.pdf")
plot(hemoTree, main = "Phylogenetic tree of NSP12 peptide (beta-coronavirus)")
dev.off()
