
<!-- this file is being included into a parent Rmd that defines all variables:
  - prefix
  - set
  - trait
-->

### Enrichment of `r trait` GWAS signal in interactors of `r set`

Boxplot:
```{r}
pvals <- get.drivers(prefix, sets, name2entrez=name2entrez, grep.cmd=grep.cmd, redo=FALSE)
pvals <- filter(pvals, SYMBOL %in% nodes(g))

ggplot(aes_string(x=set, y="ZSTAT"), data=pvals) + geom_boxplot() + geom_point(aes_string(x=set, y="ZSTAT"), data=gene_pvals[pvals[,set],]) + labs(title=trait)
```

Gene and SNP level GWAS results:
```{r}
datatable(dplyr::filter(pvals, !!(as.symbol(set))) %>% dplyr::select(SYMBOL, ZSTAT, P, variant_id, gwas_pvalue=minp))
```

Summarise the gene level P-value distribution:
```{r}
summary(dplyr::filter(pvals, !!(as.symbol(set)))[["P"]])
```

Any gene level significant GWAS hits?
```{r}
table(dplyr::filter(pvals, !!(as.symbol(set)))[["FDR"]] < 0.05)
```

Summarise the SNP level P-value distribution:
```{r}
summary(dplyr::filter(pvals, !!(as.symbol(set)))[["minp"]])
```

Any genome wide GWAS hits?
```{r}
table(dplyr::filter(pvals, !!(as.symbol(set)))[["minp"]] < 5e-8)
```

Project on network:
```{r, fig.width=20, fig.height=20}
plot.graph.with.gwas(g, pvals, virus_proteins, label.only.gwas = TRUE)
```


Gene level -log 10 p-value color legend:

```{r, fig.height=0.5, fig.width=3, echo=FALSE}
plot.color.legend(pvals)
```


